firebase-firestore/src/test/java/com/google/firebase/firestore/DocumentSnapshotTest.java

```diff
@@ -45,12 +45,13 @@ public void testEquals() {
     assertNotEquals(base, differentData);
     assertNotEquals(base, fromCache);
 
+    // Note: `base` and `differentData` have the same hash code since we no longer take document
```

Comment:
This can't be right. `equals` and `hashCode` have to be consistent or all kinds of hell breaks loose when you use these in hash sets or as keys in a map. If two objects are equal for equality purposes, the contract for hashCode requires that the hashCodes be equal too.

We can't just ignore `data` in equality. There are a few cases to consider:

  * When version=`SnapshotVersion.MIN` the documents are just made-up, so data must play a roll
  * When `DocumentState` isn't synced, the documents are potentially dirty even if they have a version

However if we have a non-`MIN` version and the `DocumentState` is `SYNCED` then we could ignore the data for equality and hashing purposes. This would preserve what you're after here which is avoiding rehydrating the data in the common case of loading a value from the remote document cache.

What do you think of implementing both equality and hashCode according to the principle above? I think that would give most (if not all) the benefit without compromising the contract of equals/hashCode.

Reply:
Hrm. This is a good comment, and I 100% agree with you, but I don't think I broke the contract here in the way that you stated. I removed `data` from hash code (I am still using document key and version). This might cause more hash collisions, but the invariant that two equal objects have the same hash code should remain intact.

Reply:
Ha! Sorry about misreading that. I didn't realize that this was about an expected collision.

An alternative phrasing that might be less confusing is:

> Note: the assertions below that hash codes of different values are not equal is not something that can be guaranteed. In particular `base` and `differentData` have a hash collision because we don't use data in the hashCode.

You might also want to call attention to this in an implementation comment in `DocumentSnapshot.hashCode()`. That way someone won't see that the value is unused in the future and then accidentally "fix" it.

Reply:
Good idea about the additional comment. I added it to `Document.hashCode()`.